aws/cloudformation-templates/apigateway.yaml

```diff
@@ -0,0 +1,657 @@
+---
+AWSTemplateFormatVersion: 2010-09-09
+
+Description: >
+    This template deploys the Retail Demo Store API Gateway
+
+Parameters:
+  ProductsServiceELBListener:
+    Type: String
+  UsersServiceELBListener:
+    Type: String
+  CartsServiceELBListener:
+    Type: String
+  OrdersServiceELBListener:
+    Type: String
+  RecommendationsServiceELBListener:
+    Type: String
+  VideosServiceELBListener:
+    Type: String
+  SearchServiceELBListener:
+    Type: String
+  LocationServiceELBListener:
+    Type: String
+  Subnets:
+    Type: String
+  VpcId:
+    Type: String
+  VpcCidr:
+    Type: String
+  ResourceBucket:
+    Type: String
+  ResourceBucketRelativePath:
+    Type: String
+  CognitoUserPoolId:
+    Type: String
+  CognitoAppClientId:
+    Type: String
+
+Resources:
+  LambdaAuthorizerRole:
+      Type: AWS::IAM::Role
+      Properties:
+        AssumeRolePolicyDocument:
+          Version: 2012-10-17
+          Statement:
+            - Effect: Allow
+              Principal:
+                Service:
+                  - lambda.amazonaws.com
+              Action:
+                - sts:AssumeRole
+        Path: /
+        ManagedPolicyArns:
+          - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+
+  LambdaAuthorizerFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      Code:
+        S3Bucket: !Ref ResourceBucket
+        S3Key: !Sub '${ResourceBucketRelativePath}aws-lambda/apigw-authorizer.zip'
+      Handler: index.handler
+      Runtime: nodejs18.x
+      Role: !GetAtt LambdaAuthorizerRole.Arn
+      MemorySize: 512
+      Timeout: 60
+      Environment:
+        Variables:
+          USER_POOL_ID: !Ref CognitoUserPoolId
+          CLIENT_ID: !Ref CognitoAppClientId
```

Comment:
I'm sure there's a reason for using a custom authorizer but curious why [API GW Cognito authorizer](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) cannot be used?

Reply:
There's a couple of reasons.  First, a Cognito Authorizer is only available for a REST API, whereas we're using a HTTP API.  Second, we need to support _guest_ access, which Cognito User Pools doesn't natively support - therefore, using a Cognito Authorizer would force you into a workaround, such as creating a specific _guest_ user